rls/src/test/java/io/grpc/rls/internal/DoNotUseDirectScheduledExecutorService.java

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.rls.internal;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+import static org.mockito.Mockito.mock;
+
+import com.google.common.base.MoreObjects;
+import io.grpc.internal.TimeProvider;
+import java.util.Comparator;
+import java.util.PriorityQueue;
+import java.util.concurrent.Delayed;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * A fake minimal implementation of {@link ScheduledExecutorService} *only* supports
+ * {@link ScheduledExecutorService#scheduleAtFixedRate(Runnable, long, long, TimeUnit)} (at most 1
+ * task is allowed) and {@link ScheduledExecutorService#schedule(Runnable, long, TimeUnit)}. It is
+ * directExecutor equivalent for {@link ScheduledExecutorService}.
+ *
+ * <p>Example:
+ * <pre>
+ * import static org.mockito.Mockito.CALLS_REAL_METHODS;
+ * import static org.mockito.Mockito.mock;
+ *
+ * private final DoNotUseDirectScheduledExecutorService fakeScheduledService =
+ *     mock(DoNotUseDirectScheduledExecutorService.class, CALLS_REAL_METHODS);
+ * </pre>
+ *
+ * <p>Note: This class is only intended to be used in this test with CALL_REAL_METHODS mock. This
+ * implementation is not thread-safe. Not safe to use elsewhere.
+ */
+abstract class DoNotUseDirectScheduledExecutorService implements ScheduledExecutorService {
+
+  private long currTimeNanos;
+  private long period;
+  private long nextRun;
+  private AtomicReference<Runnable> repeatedCommand;
+  private PriorityQueue<ScheduledRunnable> scheduledCommands;
+  private boolean initialized = false;
+
+  private synchronized void maybeInit() {
+    if (initialized) {
+      return;
+    }
+
+    initialized = true;
+    repeatedCommand = new AtomicReference<>();
+    scheduledCommands = new PriorityQueue<>(11, new ScheduledRunnableComparator());
```

Comment:
Seems every method of `DoNotUseDirectScheduledExecutorService` calls `maybeInit()`, why not just init these fields directly at their declaration and remove `maybeInit()`.

Reply:
this was due to the CALLS_REAL_METHODS doesn't initialize instance variables.

Reply:
Maybe also add a non-constructor
```
private DoNotUseDirectScheduledExecutorService() {
  throw new UnsupportedOperationException("this class is for mock only");
}
```

Btw, did you ever consider implementing `FakeClock.ScheduledExecutorImpl.scheduleAtFixedRate()`? Will there be any problem?

Reply:
the only reason i didn't implement scheduleAtFixedRate is i don't need it. it can be added if we need. also supporting multiple scheduled tasks is not hard. just don't need it at this point.